pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,223 @@
+				}
+			}
+		}
+		for _, item := range bdg.Spec.Workload.ClusterScope {
+			cWObjRef := item.ClusterScopeDownsyncObject
+			lister, found := c.listers.Get(schema.GroupVersionResource(cWObjRef.GroupVersionResource))
+			if !found {
+				return fmt.Errorf("could not get lister for gvr: %s", cWObjRef.GroupVersionResource)
+			}
+			clusterWObj, err := lister.Get(cWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := clusterWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: cWObjRef.Group, Version: cWObjRef.Version, Kind: clusterWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   cWObjRef.Resource,
+					ObjectName: cache.NewObjectName("", cWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should have status synced", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName)
+					c.singletonState.set(wObjIdentifier)
+				} else {
+					logger.V(2).Info("Singleton workload object status syncing is not driven by Binding", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName, "binding", bdg.Name)
+					c.singletonState.addIfNotExist(wObjIdentifier)
+				}
+			}
+		}
+	}
+
+	if !reconcile {
+		return nil
+	}
+
+	for wObjIdentifier := range c.singletonState.wObjSync {
+		if err := c.reconcileSingletonByWObj(ctx, wObjIdentifier); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Controller) reconcileSingletonByWObj(ctx context.Context, wObjID util.ObjectIdentifier) error {
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Reconciling singleton workload object", "resource", wObjID.Resource, "objectName", wObjID.ObjectName)
+
+	if !c.singletonState.get(wObjID) {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &wObjID, emptyStatus, c.listers, c.wdsDynClient)
+	}
+
+	list, _ := c.workStatusLister.ByNamespace("").List(labels.Everything())
+	for _, obj := range list {
+		wsRef, err := runtimeObjectToWorkStatusRef(obj)
+		if err != nil {
+			return err
+		}
+		sourceObjID := wsRef.SourceObjectIdentifier
+		if sourceObjID != wObjID {
+			continue
+		}
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return err
+		}
+		if status == nil {
+			return nil
+		}
+		logger.V(2).Info("Updating singleton status", "workstatus", wsRef.Name, "wecName", wsRef.WECName)
+		return updateObjectStatus(ctx, &wsRef.SourceObjectIdentifier, status, c.listers, c.wdsDynClient)
+	}
+	return nil
+}
+
+func (c *Controller) reconcileSingletonByBdg(ctx context.Context, ref bindingRef) error {
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Reconciling singleton status due to binding changes", "name", string(ref))
+	return c.buildSingletonStateAndOptionallyReconcile(ctx, true)
```

Comment:
What happened to the "TODO" here? It still seems to me to be problematic to do this global operation every sync of a Binding.

Reply:
Explained [here](https://github.com/kubestellar/kubestellar/pull/2285#discussion_r1709660194).